src/ape/api/projects.py

```diff
@@ -0,0 +1,259 @@
+    This is where ``ape`` will look for source files when compiling
+    the manifest for this dependency. **NOTE**: This must be the name
+    of a directory in the project.
+    """
+
+    _data_folder: Path
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__} name='{self.name}'>"
+
+    @property
+    @abstractmethod
+    def version_id(self) -> str:
+        """
+        The ID to use as the sub-directory in the download cache.
+        Most often, this is either a version number or a branch name.
+        """
+
+    @property
+    def _target_manifest_cache_file(self) -> Path:
+        version_id = self.version_id
+        if isinstance(
+            version_util.parse(version_id), version_util.Version
+        ) and not version_id.startswith("v"):
+            version_id = f"v{version_id}"
+
+        name = self.name
+        return self._data_folder / "packages" / name / version_id / f"{name}.json"
+
+    @abstractmethod
+    def extract_manifest(self) -> PackageManifest:
+        """
+        Create a :class:`~ape.api.projects.ProjectAPI` implementation,
+        presumably by downloading and compiling the dependency.
+
+        Implementations may use ``self.project_manager`` to call method
+        :meth:`~ape.managers.project.ProjectManager.get_project`
+        to dynamically get the correct :class:`~ape.api.projects.ProjectAPI`.
+        based on the project's structure.
+
+        Returns:
+            :class:`~ape.api.projects.ProjectAPI`
+        """
+
+    @property
+    def cached_manifest(self) -> Optional[PackageManifest]:
+        """
+        The manifest from the ``.ape/packages/<dependency-name>/<version-id>``
+        if it exists and is valid.
+        """
+
+        return _load_manifest_from_file(self._target_manifest_cache_file)
+
+    def _extract_local_manifest(self, project_path: Path):
+        contracts_folder = project_path / self.contracts_folder
+        project = self.project_manager.get_project(
+            project_path,
+            contracts_folder=contracts_folder,
+            name=self.name,
+            version=self.version,
+        )
+        sources = [
+            s
+            for s in get_all_files_in_directory(project.contracts_folder)
+            if s.name.lower() not in ("package.json", "package-lock.json")
+        ]
+        project_manifest = project.create_manifest(file_paths=sources)
+
+        # Cache the manifest for future use outside of this tempdir.
+        self._target_manifest_cache_file.parent.mkdir(exist_ok=True, parents=True)
+        self._target_manifest_cache_file.write_text(json.dumps(project_manifest.dict()))
+
+        return project_manifest
+
+
+def _load_manifest_from_file(file_path: Path) -> Optional[PackageManifest]:
+    if not file_path.exists():
+        return None
+
+    def handle_corrupt_manifest():
+        logger.warning(f"Existing manifest file '{file_path}' corrupted (re-building).")
+        file_path.unlink()
```

Comment:
I feel like the caller of this should handle deletion, so we don't end up with a weird silent bug

Reply:
Don